### ISSUE TITLE
Badge Earning Issue, or Limitation in Current Code Structure?

### ISSUE THREAD

**Issue by adamapollo** [opencredit/badgeos#449]
Hey Guys,

This is the only place I've seen to submit an inquiry about the BadgeOS Core feature set and functionality. This may be a bug, or it may be a critical component of the core feature set that has not been implemented.

Let's say you have a series of courses that each have a Badge. Users are taking the courses, and earning Badges. You then determine that a group of courses should earn a Level Badge, one that is earned by completing the specific Achievements of each of the individual Badges in that level. Alternatively, it could be earned by completing the series of Learndash courses that should cumulatively earn the Level Badge.

In either scenario, if the Level Badge is created AFTER any Users have earned any of the Achievements that compose the STEPS necessary to earn that Level Badge, then they will NOT earn the Level Badge upon completion of the Steps completed. They will be recorded ONLY for Steps completed AFTER the new Level Badge was created. This is because the system only seems to be registering Steps completed after a Badge is created, and not checking the DB for previous Achievements. 

Is this a limitation of the current code, or an Issue that has not been addressed?

It seems to me that no matter when you create a Badge, the BadgeOS should be doing a DB check to see what Steps have been completed, or Achievements Earned, after ANY new Achievement is earned or Step is completed. This would enable a scalable Badge Earning model where Admins can continue to create New Badges and Achievements that incorporate ones made previously.

I thought this was the built in functionality, but I've tested it both by using Earned Achievements as Steps and also Completed Learndash Courses as Steps, and neither route is working. 

I'm now faced with a much greater challenge, issuing a Level Badge to hundreds of users who are completing the course series that is leading to that Level, since I didn't create the Level Badge until half-way through our 9 week course.

Please let me know if there is a workaround, solution, or an easy way to:

1. Set up a Badge to incorporate previously earned Achievements.
2. Bulk Add a Badge to a Series of Users who have previously earned Achievements. (How do I find these users besides opening each of their User editors in wp-admin in turn???)
3. Bulk remove a Badge from a series of users. (What if users have earned Badges that were not set up properly initially? How do we remove them in Bulk?)

Thank you for your timely reply, and for an otherwise awesome Achievement system!

Highest Regards,
Adam Apollo

Superluminal Systems

:::

Screenshots:

In Level Badge, only the Achievements completed after it was created are listed:
![image](https://cloud.githubusercontent.com/assets/1397136/7980660/1b23ac52-0a5e-11e5-8528-8d1d890c4a28.png)

Yet the user has earned the prior Achievements when looking at the individual Badges that make up that Level:
![image](https://cloud.githubusercontent.com/assets/1397136/7980668/32643468-0a5e-11e5-934b-e998a9e66f9a.png)

**Comment by tw2113** [comment]
Yes, from everything I've seen, it's a limitation, and I would love to offer a way to re-validate existing achievements with what's available to a given user and possibly all. However, I can't guarantee it'd get in. The ticket for this idea is #230 

I can understand both sides of the coin for why this type of thing should or shouldn't be handled in such a way, for what it's worth.

Our user achievement management is also one of the weaker areas at this point, and I know I have a lot of tickets on GitHub here related to it.

**Comment by adamapollo** [comment]
Thanks Michael. Any suggestions on approaching this now *very* challenging situation for the time being?

**Comment by tw2113** [comment]
Sadly I don't have any workaround suggestions available that are tested. Best I have is the manual process or seeing if the user re-doing any/all of the associated steps to see if that triggers it as well.

**Comment by adamapollo** [comment]
Michael, I think I've got a solution for my situation that may also work for others:

1. Add a new Lesson or Topic page to a Learndash course that is connected to a BadgeOS Achievement. 
2. Request the Users now mark that Lesson or Topic complete (we're adding a media page with all the videos from the course).
3. This should re-trigger "course complete" from Learndash, which should then trigger the Step for BadgeOS, correct?

Also, in trying to address this issue, I created some test Badges with other steps. I now have what looks like many duplicate steps in the list when I go to "Award Step" or similar from the User Page. Also, all Steps have a default icon, and I don't have an Achievement Type called steps, but they all show up in the media library... 

Can I create an Achievement Type called "Steps" to have them show up and manage them? I can't figure out how to manage steps otherwise, and I'm concerned about deleting the repetitive icons from the media library... 

Your help is immensely appreciated.

**Comment by tw2113** [comment]
Without going too deep into thought at the moment, we already create a post type of "steps" and keep it visually hidden, but still used. Creating your own of the same name, I believe, would overwrite ours.

Frequently, when I've debugged issues for people who say they were seeing duplicate awards in the user admin, it was simply a case of the step post type post and their achievement post type post being named the same thing. 

I have a bug fix that I can point out, if you want to apply it, that gets rid of the repetitive media library graphics, however, since steps shouldn't really be shown to users in most cases, getting rid of the graphic shouldn't be a huge deal anyway.

**Comment by adamapollo** [comment]
Yes, creating the "Steps" post type on staging screwed up the system, and caused any new steps added to show up as new Achievement Types.

I copied to staging again, and this time just added a lesson to each course in the series that earns the level badge, and then tested. A few of the steps did NOT register this time, but once I deleted the steps and recreated them, they registered normally and the level badge was earned.

Just for clarity, you're saying that the image that shows up in the media library repetitively has no direct relation to the steps themselves, so it's safe to delete them without affecting the steps?

Also, how would one change the default image for the steps?

Thanks for all your help and prompt responses. I've managed to do a lot with BadgeOS so far and am working to refine my process. A little more information about "order of operations" would be really useful in the FAQ or Tutorials for new users so they don't run into these problems.

**Comment by tw2113** [comment]
"Just for clarity, you're saying that the image that shows up in the media library repetitively has no direct relation to the steps themselves, so it's safe to delete them without affecting the steps?"
The attachment ID is set for the step's featured image field, but that's update-able as necessary. I believe we have a filter on it available that you point to a static image somewhere.

```
apply_filters( 'badgeos_default_achievement_post_thumbnail', 'https://credlyapp.s3.amazonaws.com/badges/af2e834c1e23ab30f1d672579d61c25a_15.png' );
```

If there's no achievement graphic already associated, with the achievement, it'll descend into some if statements and grab the file filtered above and if it can't find the file already in the media library, it'll actually upload the fetched file.

See it all at https://github.com/tw2113/badgeos/blob/alpha/includes/achievement-functions.php#L714

You pointed out to me, accidentally, that my duplicate images in library won't apply to custom defaults, so I have a new bug to file.

**Comment by adamapollo** [comment]
Hey Michael, I'm in the process of upgrading an Academy, and we have come up against this problem again. 

Users need to complete six Learndash courses in order to earn a "Series Badge." They also earn a Badge for each course completion, so there are Steps for each course completion. 

Previously, the courses had to be completed in a specific order. We set the "Series Badge" to be issued along with the course Badge when the user finished the last course in the series.

However, now the courses can be taken in any order. We need to register that users have completed all the Steps that are already created, yet I don't see an option in BadgeOS to select existing Steps in order to release a Badge, only create a new series of Steps. This makes a duplicate set of Steps that register for the same thing: users completing the six courses in the Series.  

The problem gets bigger when you consider that some users have already completed some courses in the series, but not others. If they complete the remaining courses now, they will not trigger these new Steps applied to courses they have already done, so they will not earn the Series Badge...

As far as I can figure out, I need one of the following options to work:

1. Selecting EXISTING Steps as the triggers for a new Badge.
2. Some way to run a "DB Check and Update" to see if users have completed steps in a course or completed the Learndash course and have their Steps updated in BadgeOS.
3. BadgeOS simply recognizing that if a user already has certain achievements, then if a new achievement is added that requires those achievements, they will actually earn that achievement when the last of those existing achievements are completed, or immediately if they already have all the achievement steps required.

I'm open to other options or suggestions here again, as this is now a critical issue and I haven't found another workaround.
